### DISCUSSION TITLE
Stuck on loop

### DISCUSSION THREAD

**Discussion by VijayDamodharan** [Part-II-Computational-Physics/.github#17] [General]
For Core Task 2, I have the following code defined:

```
def Lax_Wendroff(v, n_points_space, dx, dt):

    def step(f):
        l = v/2*dx
        p = (v**2 * dt)/(2 * dx**2) 

        df_dt = f.copy()
        for i in range(n_points_space):
            if i == n_points_space - 1:  # cannot do f[i+1] term for the last term as it'll give index error
                df_dt[i] = -l * (f[0] - f[i - 1]) + p * (f[0] - 2 * f[i] + f[i - 1])
            else:
                df_dt[i] = -l * (f[i + 1] - f[i - 1]) + p * (f[i + 1] - 2 * f[i] + f[i - 1])
        f[:] += dt * df_dt
    
    return step
```

However, I am having 2 problems. The animation is not working (it is stuck on just one frame) and when I run the code for error testing, it seems to get stuck on a loop and never finishes execution. Any ideas why?

**Comment by NathanMagnan** [comment]
I'm not sure for the animation, did it work in the explicit upwind case?

For the error code, it can take some time because it needs to solve a few complicated problems. I would try and put a print on every iteration of the loop, to monitor the progress. If it's stuck on the first iteration then you have a problem because this one should be quick. but maybe it just takes a long time on iterations 19 and 20.

**Reply by VijayDamodharan** [reply]
You were right. The errors just took a long time to calculate, but printing out each iteration showed it was working. Is taking 13 mins to calculate all 20 iterations normal?

The animation for Upwind didn't run when I was using VSCode (despite removing the magic method at the start, as experienced by some other people). So I copy-pasted my code into PyCharm and it worked fine. But when I copy-pasted my above code for Lax_Wendroff into Pycharm, it still doesn't animate properly. Not sure if there's something wrong with my code..?

**Reply by NathanMagnan** [reply]
13m is very long, it takes about 50s on my machine (which is not particularly high-end). This is probably related to how efficient your particular implementation of the Lax-Wendroff algorithm is. But code optimization is not the topic of this problem set, so let's not worry about it.

If the animation works for the explixit upwind method but not for Lax-Wendroff, the problem must be inside the Lax-Wendroff function. Possible guesses are:
- A compilation error being hidden by jupyter (I'm thinking of the i = 0 case, you access the index -1 of the list, which might or might not work)
- Python's sometime obscure way of handling passage by reference (I'm thinking of the `df_dt = f.copy()` line, you could try replacing it with `df_dt = np.zeros(n_points_space)`

Although, if the error measurement code works for the Lax-Wendroff scheme, it means your Lax-Wendroff scheme is correct, so it's even more confusing. Maybe make sure you didn't modify anything in the animation function?

If neither guess works, then I don't really know. 
Sometime deleting one's work and restarting from scratch works

**Reply by VijayDamodharan** [reply]
I reran the code using `df_dt = np.zeros(n_points_space` as suggested.

Yes, the error measurement code ran completely fine and gave the expected graphical output. I can paste my animation function below so you can see if there's anything wrong:

```
# initial condition
X = np.linspace(x_min, x_max, n_points_space, endpoint=True)
f0 = gaussian(X)
f = f0.copy()

# scheme
step = Lax_Wendroff(v, n_points_space, dx, dt)

# preparation of the animation
fig = plt.figure()
ax = fig.gca()

ax.set_xlim(x_min, x_max)
ax.set_ylim(-0.1, 1.5)

ax.set_xlabel(r'$x$')
ax.set_ylabel(r'$f(x, t)$')

show_f0, = ax.plot(X, f0, 'b--', label="f0")
show_ft, = ax.plot([], [], 'b', lw=1, ms=2, label="f")
time = ax.annotate(0, xy=(3, 1.4), xytext=(3, 1.4))


# integration
def animate(i):
    step(f)
    print(f"{i} frame")
    show_ft.set_data(X, f)

    time.set_text("t / T = %.2f" % (v * i * dt / (x_max - x_min)))

    return show_f0, show_ft


# creation of the animation
anim = animation.FuncAnimation(fig, animate, 10 ** 5, interval=dt * 10, blit=True)
# anim.save("Core task 2 animation output.gif", writer='ffmpeg')

# show the results
plt.show()
```

The `print(f"{i} frame")` is just to check that the frame were running through properly. It did print out the different frame numbers, so I'm not sure what is going wrong.
